File: apps/cli/src/services/state.ts
import * as Effect from 'effect/Effect';
import { join } from 'path';
import { homedir } from 'os';
import type { SessionStatus } from '../domain/session.ts';

const STATE_DIR = join(homedir(), '.config', 'agent-dash');
const STATE_FILE = join(STATE_DIR, 'state.json');

type PersistedState = {
	unreadPaneIds: Array<string>;
	prevStatusMap: Record<string, SessionStatus>;
};

export const loadState = Effect.gen(function* () {
	const raw = yield* Effect.tryPromise(() => Bun.file(STATE_FILE).text()).pipe(
		Effect.catchAll(() => Effect.succeed(null)),
	);

	if (raw === null) {
		return {
			unreadPaneIds: new Set<string>(),
			prevStatusMap: new Map<string, SessionStatus>(),
		};
	}

	const parsed = yield* Effect.try(() => JSON.parse(raw) as PersistedState).pipe(
		Effect.catchAll(() => Effect.succeed(null)),
	);

	if (parsed === null) {
		return {
			unreadPaneIds: new Set<string>(),
			prevStatusMap: new Map<string, SessionStatus>(),
		};
	}

	return {
		unreadPaneIds: new Set(parsed.unreadPaneIds),
		prevStatusMap: new Map(Object.entries(parsed.prevStatusMap)) as Map<string, SessionStatus>,
	};
});

export function saveState(
	unreadPaneIds: Set<string>,
	prevStatusMap: Map<string, SessionStatus>,
) {
	return Effect.gen(function* () {
		const data: PersistedState = {
			unreadPaneIds: Array.from(unreadPaneIds),
			prevStatusMap: Object.fromEntries(prevStatusMap),
		};

		yield* Effect.tryPromise(async () => {
			const { mkdir } = await import('fs/promises');
			await mkdir(STATE_DIR, { recursive: true });
			await Bun.write(STATE_FILE, JSON.stringify(data));
		}).pipe(Effect.catchAll(() => Effect.void));
	});
}

File: apps/cli-rs/src/state.rs
use crate::session::SessionStatus;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PersistedState {
    unread_pane_ids: Vec<String>,
    prev_status_map: HashMap<String, SessionStatus>,
}

fn state_dir() -> PathBuf {
    dirs::home_dir()
        .expect("home directory not found")
        .join(".config/agent-dash")
}

fn state_path() -> PathBuf {
    state_dir().join("state.json")
}

pub struct LoadedState {
    pub unread_pane_ids: HashSet<String>,
    pub prev_status_map: HashMap<String, SessionStatus>,
}

pub fn load_state() -> LoadedState {
    let path = state_path();
    let content = match std::fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => {
            return LoadedState {
                unread_pane_ids: HashSet::new(),
                prev_status_map: HashMap::new(),
            };
        }
    };
    match serde_json::from_str::<PersistedState>(&content) {
        Ok(parsed) => LoadedState {
            unread_pane_ids: parsed.unread_pane_ids.into_iter().collect(),
            prev_status_map: parsed.prev_status_map,
        },
        Err(_) => LoadedState {
            unread_pane_ids: HashSet::new(),
            prev_status_map: HashMap::new(),
        },
    }
}

pub fn save_state(unread_pane_ids: &HashSet<String>, prev_status_map: &HashMap<String, SessionStatus>) {
    let data = PersistedState {
        unread_pane_ids: unread_pane_ids.iter().cloned().collect(),
        prev_status_map: prev_status_map.clone(),
    };
    let dir = state_dir();
    let _ = std::fs::create_dir_all(&dir);
    let _ = std::fs::write(state_path(), serde_json::to_string(&data).unwrap_or_default());
}
